beyond-all-reason/Beyond-All-Reason
# Gridmenu: Optional "build mode" for factories

### Work done
This is the first piece of work for a long-term project of revamping factory behavior with gridmenu, and subsequently overhauling grid keys with the new freedom that that will allow.

- Factory building hotkeys are no longer "always on", and need to be enabled by activating "build mode"
- Build mode is activated by pressing any of the category keys while in a lab, ZXCV
- Build mode will remain active when build hotkeys are used
- Build mode can be cancelled by pressing (and releasing) shift, changing unit selection, or pressing esc.
- Hotkey hints on units will be hidden when build mode is disabled
- Con and structure behavior is unchanged, this only affects labs.

All of this behavior is behind a new option.

![image](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/3493638/b5786513-a21c-4060-865e-8183e8038266)

#### Test steps
It's hard to write detailed test steps for a feature like this. Try to switch between labs and test out the new behavior, and ensure there are no issues with it being enabled when it shouldn't be.

The behavior of shift (as a modifier for unit count in labs) is of particular interest, this may be a source of bugs or unintuitive behavior.

### Screenshots
![spring_LvtGbPqlB8](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/3493638/357cd3d0-f895-4de4-b484-1a2852381ca3)

![spring_LemZBS7wr4](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/3493638/e7e979fb-d7d8-4e5d-adc8-f271d6c871ef)
